### ISSUE TITLE
Incompatibility with Futurepack

### ISSUE THREAD

**Issue by Mordna** [SpongePowered/SpongeForge#1999]
**I am currently running**
- SpongeForge version: 1.12.2-2586-7.1.0-BETA-2899
- Forge version: 14.23.2.2611 (also tried matching version)
- Java version: java-8-openjdk 8.u144
- Operating System: Linux 4.15.1

- Plugins/Mods: Futurepack 1.12.2-26.3.88

**Issue Description**
After generating a new world, moving a bit around and the going in a TechDungeon or exploring the world (takes longer) a "runaway phase" is detected and one has to terminate the server manually. After restarting the problem starts immediately. Sometimes it takes up to half an hour, sometimes it happens after a few seconds.
It is (obviously) faster if one changes the chance of a TechDungeon to a higher value.
It wont crash while not using SpongeForge

Fml-junk-earlystartup.log: https://gist.github.com/Mordna/ff619ae5aeef426f6bd04fa7808a100c
Other logs do not contain any errors/warnings/fails or other abnormalities.

I also created [a issue](https://github.com/mcenderdragon/Futurepack-API/issues/372) on the github page of the mod, but it should be SpongeForge's fault.

EDIT:
After running the mod with spongeforge development workspace, the interesting lines are those:
```
- Phase: PluginPhase{}{ScheduledTaskPhaseState}                                                 
     Context:                                                                                      
     - Owner: null                                                                                 
     - Notifier: null                                                                              
     - Source: java.util.concurrent.Executors$RunnableAdapter@34653297                             
     - CapturedBlocks: CapturedBlocksSupplier{Captured=0}                                          
     - BlockItemDrops: BlockItemDropsSupplier{Captured=0}                                          
     - BlockItemEntityDrops: BlockItemEntityDropsSupplier{Captured=0}                              
     - CapturedItems: CapturedItemsSupplier{Captured=0}                                            
     - CapturedEntities: CapturedEntitiesSupplier{Captured=0}                                      
     - CapturedItemStack: CapturedItemStackSupplier{Captured=0}                                    
     - EntityItemDrops: null                                                                       
     - EntityItemEntityDrops: null                                                                 
     - BlockEntitySpawns: CapturedBlockEntitySpawnSupplier{Captured=0}                             
     - CapturedBlockPosition: null                                                                 
   Printing stack trace :                                                                          
 java.lang.Exception: Stack trace                                                                  
     org.spongepowered.common.event.tracking.PhaseTracker.printRunawayPhase(PhaseTracker.java:332) 
     org.spongepowered.common.event.tracking.PhaseTracker.switchToPhase(PhaseTracker.java:166)     
     org.spongepowered.common.event.tracking.PhaseContext.buildAndSwitch(PhaseContext.java:188)    
     net.minecraft.server.MinecraftServer.redirect$onCall$zje000(MinecraftServer.java:4175)        
     net.minecraft.server.MinecraftServer.callFromMainThread(MinecraftServer.java:1536)            
     net.minecraft.server.MinecraftServer.addScheduledTask(MinecraftServer.java:1548)              
     net.minecraft.world.WorldServer.addScheduledTask(WorldServer.java:1427)                       
     futurepack.common.block.TileEntityDungeonCore.run(TileEntityDungeonCore.java:128)
```

Remapped snippet of repeating runaway phase: https://gist.github.com/ryantheleach/14b6097cc9fc41223d558657882b3f1e

**Comment by ryantheleach** [comment]
```
net.minecraft.server.MinecraftServer.redirect$onCall$zjf000(MinecraftServer.java:3993)
net.minecraft.server.MinecraftServer.callFromMainThread(MinecraftServer.java:1381)
net.minecraft.server.MinecraftServer.addScheduledTask(MinecraftServer.java:1393)
net.minecraft.world.WorldServer.addScheduledTask(WorldServer.java:1334)
futurepack.common.block.TileEntityDungeonCore.run(TileEntityDungeonCore.java:128)
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
net.minecraft.server.MinecraftServer.redirect$onCall$zjf000(MinecraftServer.java:3994)
net.minecraft.server.MinecraftServer.callFromMainThread(MinecraftServer.java:1381)
net.minecraft.server.MinecraftServer.addScheduledTask(MinecraftServer.java:1393)
net.minecraft.world.WorldServer.addScheduledTask(WorldServer.java:1334)
futurepack.common.block.TileEntityDungeonCore.run(TileEntityDungeonCore.java:128)
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
```
TileEntityDungeonCore.run: 
https://gist.github.com/ryantheleach/14b6097cc9fc41223d558657882b3f1e#file-tileentitydungeoncore-java-L91-L105

It looks like we are calling the callable immediately, where as a default minecraft only calls the callable immediately if (!this.isCallingFromMinecraftThread() && !this.isServerStopped())) == false

https://github.com/SpongePowered/SpongeCommon/blob/09040e65251084c0081c2d5e3dd7a5176be41a41/src/main/java/org/spongepowered/common/mixin/core/server/MixinMinecraftServer.java#L796

@Aaron1011 git blamed you as the creator of this code, was the purpose of this only to add in the phase tracking? Was this originally written against 1.12? I'm weary that our code is calling this potentially on the wrong thread as well?

**Comment by Mordna** [comment]
If I debug forge server while using the mod and watch the expression after
`if (!.this.isCallingFromMinecraftThread() && !this.isServerStopped())`
as you mentionend in MinecraftServer.class, Futurepacks' task are executed there.
Tasks include: DungeonGeneratorBase$BakedDungeon

If I do the same in Spongeforge, only adding/replacing 
```
Object value = null;
        if(!this.isMainThread() && this.isServerRunning()) {        //Check for different thread
            ListenableFutureTask<?> listenablefuturetask = ListenableFutureTask.create(callable);
            
            //The future task queue will be updated on updateTimeLightAndEntities();
            synchronized (this.futureTaskQueue)
            {
                this.futureTaskQueue.add(listenablefuturetask);
                value = listenablefuturetask;
            }
        } else {
            try (BasicPluginContext context = PluginPhase.State.SCHEDULED_TASK.createPhaseContext()
                    .source(callable)
                    .buildAndSwitch()) {
                value = Futures.immediateFuture(callable.call());
            } catch (Exception e) {
                throw e;
            }
        }
        return value;
```
in ```public Object onCall(Callable<?> callable) throws Exception...``` in MixinMinecraftServer.class
And adding
```
 //Shadow Future task queue to add tasks, needs to be initialized
@Shadow public final Queue< FutureTask<? >> futureTaskQueue = Queues.newArrayDeque();
``` 
to the variables in MixinMinecraftServer.class to avoid build errors, it never reaches the code inside 
```if() {}```, while the other tasks are handled perfectly fine by the current system. Since the server was running it seams that you were right with the wrong thread for the caller, since ```isRunning() is equal to !isServerStopped```. 

Maybe that helps to find the bug if all above is right (maybe somebody other can verify this).

EDIT:
Did another test and in forge, addScheduledTask is <b>not</b> executed in the server thread while in spongeforge it is executed <b>in</b> the server thread. That should not be, even if it sometimes (up to half an hour running around -> generating new tasks) works there should be a better solution.

**Comment by ProsperCraft** [comment]
Same problem on my end.

**Comment by mcenderdragon** [comment]
I often use addSheduled task to get/set BLockstates of the world from other threads, so please keep that in mind to prevent something like #2024

**Comment by Zidane** [comment]
@Mordna 

You code sample is incorrect. We redirect callable.call so we are inside the:

``` java
Futures.immediateFuture(<PhaseTracker Block Here>
```

**Comment by liach** [comment]
@Mordna Does this still happen in futurepack 26.3.123? The related code changed.

**Comment by Mordna** [comment]
@liach It does not happen in 26.3.152.
Is it normal that typing just 'sponge' in the server command line results in:
`Invalid subcommand state -- no more than one mapping may be provided for child arg child0`
while typing 'sponge foo' displays valid subcommands?
Using forge-2611, sponge 2953, 1.12.2

@Zidane I am sorry, I only thought of going backwards and tested the behavior against forge, but i did not check Futures.immedateFuture or the redirection of callable.call. So I simply wrote what I found without much further testing.
